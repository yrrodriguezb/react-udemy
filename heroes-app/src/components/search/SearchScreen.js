import React, { useMemo } from 'react'
import queryString from 'query-string'
import { HeroCard } from '../hero/HeroCard';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { useLocation, useNavigate } from 'react-router-dom';

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);

  const [ formValues, handleInputChange ] = useForm({
    searchText: q
  });

  const { searchText } = formValues;
  const heroes = useMemo(() => getHeroesByName(q), [ q ]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchText }`)
  }

  return (
    <>
      <h1>Busqueda</h1>
      <hr/>

      <div className="row">
        <div className="col-5">
          <h4>Buscar</h4>
          <hr />

          <form onSubmit={ handleSearch }>
            <input 
              type="text"
              placeholder="Buscar un héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ handleInputChange }
            />

            <button
              type='submit'
              className='btn btn-outline-primary mt-3'
            >
              Buscar
            </button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Resultados: { heroes.length }</h4>
          <hr />
          {
            (q === '') 
              ? <div className='alert alert-info'>Buscar un héroe</div>
              : (heroes.length === 0)
                && <div className='alert alert-danger'>No hay resultados para: { q }</div> 
          }
          {
            heroes.map(hero => (
              <HeroCard key={ hero.id } hero={ hero }/>
            ))
          }
        </div>
      </div>  
    </>
  )
}
