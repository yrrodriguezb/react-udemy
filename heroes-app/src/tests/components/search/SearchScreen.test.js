import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';
import '@testing-library/jest-dom'


// Es importante que la variable inicie con mock
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate // Sobreesribimos el hook
}))

describe('Pruebas en <SearchComponente />', () => {
  test('Debe mostrarse correctamente', () => {
    // Es necesario el MemoryRouter para evitar el siguiente error:
    //  useNavigate() may be used only in the context of a <Router> component.

    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search'] }>
        <SearchScreen />
      </MemoryRouter>
    ) 

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un héroe');
  });

  test('Debe mostrar a batman y el input con el valor del querystring', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=batman'] }>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe('batman');
  });

  test('Debe mostrar el mensaje que indica que no se encontrarón datos con el parametro de busqueda', () => {
    const query = 'batman_123_yrr'
    const wrapper = mount(
      <MemoryRouter initialEntries={ [`/search?q=${query}`] }>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find("HeroCard").length).toBe(0);
    expect(wrapper.find(".alert.alert-danger").exists()).toBe(true);
    expect(wrapper.find(".alert.alert-danger").text().trim()).toBe(`No hay resultados para: ${query}`);
  });

  test('Debe llamar el navigate a la nueva URL', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search'] }>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', { 
      target: { 
        name: 'searchText', 
        value:'batman' 
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {}
    });

    expect(mockNavigate).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith("?q=batman");
  })
})