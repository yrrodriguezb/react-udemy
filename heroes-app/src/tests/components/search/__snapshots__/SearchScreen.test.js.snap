// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchComponente /> Debe mostrar a batman y el input con el valor del querystring 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados: 
        1
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/img/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
          <div />
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchComponente /> Debe mostrarse correctamente 1`] = `
Array [
  <h1>
    Busqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados: 
        0
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
